sdk/internal/testframework/request_matcher.go

```diff
@@ -0,0 +1,176 @@
+// +build go1.13
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package testframework
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"github.com/dnaeon/go-vcr/cassette"
+)
+
+type RequestMatcher struct {
+	context        TestContext
+	IgnoredHeaders map[string]*string
+	bodyMatcher    StringMatcher
+	urlMatcher     StringMatcher
+	methodMatcher  StringMatcher
+}
+
+type StringMatcher func(reqVal string, recVal string) bool
+type matcherWrapper func(matcher StringMatcher, testContext TestContext) bool
+
+var ignoredHeaders = map[string]*string{
+	"Date":                   nil,
+	"X-Ms-Date":              nil,
+	"x-ms-date":              nil,
+	"x-ms-client-request-id": nil,
+	"User-Agent":             nil,
+	"Request-Id":             nil,
+	"traceparent":            nil,
+	"Authorization":          nil,
+}
+
+const (
+	recordingHeaderMissing = "Test recording headers do not match. Header '%s' is present in request but not in recording."
+	requestHeaderMissing   = "Test recording headers do not match. Header '%s' is present in recording but not in request."
+	headerValuesMismatch   = "Test recording header '%s' does not match. request: %s, recording: %s"
+	methodMismatch         = "Test recording methods do not match. request: %s, recording: %s"
+	urlMismatch            = "Test recording URLs do not match. request: %s, recording: %s"
+	bodiesMismatch         = "Test recording bodies do not match.\nrequest: %s\nrecording: %s"
+)
+
+func DefaultMatcher(testContext TestContext) *RequestMatcher {
```

Comment:
Can we also not export this?  I see there's an exported `Matcher` on the `Recording`, is the expectation that custom matching would be obtained by manipulating this field?

Reply:
The `Matcher` field on `Recording is an instance of this type. The intended means for manipulating it would be:
- Modify `IgnoredHeaders`
- Set a custom matcher by calling `SetBodyMatcher`, `SetURLMatcher`, or `SetMethodMatcher`

Reply:
It would probably be more idiomatic to name this `NewRequestMatcher`. I initially named it this to convey that it is pre-configured with the default behavior and doesn't require calls to any of the `Set*Matcher` methods if you prefer the default behavior. I'll add a comment.